File: src/main/java/xyz/upperlevel/hgame/event/SimpleEventListener.java
package xyz.upperlevel.hgame.event;

import lombok.EqualsAndHashCode;

import java.util.function.Consumer;

@EqualsAndHashCode(callSuper = true)
public class SimpleEventListener<E extends Event> extends EventListener<E> {
    private final Consumer<E> consumer;

    public SimpleEventListener(Class<E> clazz, byte priority, Consumer<E> consumer) {
        super(clazz, priority);
        this.consumer = consumer;
    }

    public SimpleEventListener(Class<E> clazz, Consumer<E> consumer) {
        this(clazz, EventPriority.NORMAL, consumer);
    }

    @Override
    public void call(E event) {
        consumer.accept(event);
    }
}

File: src/main/kotlin/xyz/upperlevel/hgame/event/SimpleEventListener.kt
package xyz.upperlevel.hgame.event

class SimpleEventListener<E : Event>(clazz: Class<E>,
                                     private val consumer: (E) -> Unit,
                                     priority: Byte = EventPriority.NORMAL) : EventListener<E>(clazz, priority) {
    override fun call(event: E) {
        consumer(event)
    }

    override fun equals(other: Any?): Boolean {
        return super.equals(other) and
                (this.consumer == (other as SimpleEventListener<*>).consumer)
    }

    override fun hashCode(): Int {
        return super.hashCode() + 59 * consumer.hashCode()
    }
}
